src/database/arangodb/aql-generator.ts

```diff
@@ -1399,153 +1418,842 @@ register(FollowEdgeQueryNode, (node, context) => {
+    }
+
+    // this is very similar to processTraversalWithOnlyRelationSegmentsNoList(),
+    // but instead of using the rootVar in mapping, we use the field traversal result
+
+    const rootVar = aql.variable(`root`);
+    const forStatementsFrag = getRelationTraversalForStatements({
+        node,
+        innermostItemVar: rootVar,
+        preserveNullValues: node.preserveNullValues,
+        context,
+    });
 
+    const fieldTraversalFrag = getFieldTraversalFragment({
+        segments: node.fieldSegments,
+        sourceFrag: rootVar,
+    });
+    const innerContext = context.introduceVariableAlias(node.itemVariable, fieldTraversalFrag);
+
+    return aqlExt.parenthesizeObject(
+        forStatementsFrag,
+        aql`RETURN ${node.innerNode ? processNode(node.innerNode, innerContext) : fieldTraversalFrag}`,
+    );
+}
+
+/**
+ * Produces:
+ *
+ *     FOR node1 IN OUTBOUND source_hop1
+ *     FOR node2 in OUTBOUND hop1_hop2
+ *     FOR root IN OUTBOUND hop2_target
+ *     FOR item IN root.fieldSegment1[*].fieldSegment2[** FILTER filterExpr(CURRENT)][*]
+ *     SORT item.sortValues[0]
+ *     LIMIT skip, maxCount
+ *     RETURN innerNode(item)
+ *
+ * or, if preserveNullValues is true and hop2_target is a to-1 relation:
+ *
+ *     FOR node1 IN OUTBOUND source_hop1
+ *     FOR node2 in OUTBOUND hop1_hop2
+ *     LET root = FIRST(FOR node3 IN OUTBOUND hop2_target RETURN node3)
+ *     FOR item IN root.fieldSegment1[*].fieldSegment2[** FILTER filterExpr(CURRENT)][*]
+ *     SORT item.sortValues[0]
+ *     LIMIT skip, maxCount
+ *     RETURN innerNode(item)
+ */
+function processTraversalWithRelationAndListFieldSegmentsUsingSubquery(
+    node: TraversalQueryNode,
+    context: QueryContext,
+) {
+    if (!node.relationSegments.length || !node.fieldSegments) {
+        throw new Error(`Expected both relation and field segments`);
+    }
+    if (!node.fieldSegments.some((s) => s.isListSegment)) {
+        throw new Error(`Expected at least one list field segment`);
+    }
+    if (!node.resultIsList) {
+        throw new Error(`Cannot have sort on non-list traversal`);
+    }
+
+    // this will hold the node of the innermost relation traversal
+    const rootVar = aql.variable('root');
+
+    const forStatementsFrag = getRelationTraversalForStatements({
+        node,
+        innermostItemVar: rootVar,
+        preserveNullValues: node.preserveNullValues,
+        context,
+    });
+
+    const filterNode = node.filterNode;
+    const innerFilterFrag = filterNode
+        ? (itemFrag: AQLFragment) => {
+              // don't provide rootEntityVariable
+              // (if we want to filter on root, it should happen outside already)
+              const innerContext = context.introduceVariableAlias(node.itemVariable, itemFrag);
+              return processNode(filterNode, innerContext);
+          }
+        : undefined;
+
+    const outerMapFrag = getFieldTraversalFragment({
```

Comment:
nit: I think just naming it  `fieldTraversalFrag` would be easier to understand.